### PR TITLE
Adding testing

### PR DESCRIPTION
**Pull request by ssadhu-sl**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

1. Added Unit & Integration Testing
2. Added Black and flake and test-cov (to make sure cov is above 90% all time)
3. Few tweaks in the code to accommodate testing

## Related Tickets & Documents

- Ticket Link: https://github.com/orgs/digital-land/projects/9?pane=issue&itemId=83945113&issue=digital-land%7Ctechnical-documentation%7C106

## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes, a note
on the devices and browsers this has been tested on, as well as any relevant
images for UI changes._

## Added/updated tests?
_We encourage you to keep the code coverage percentage at 80% and above._

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [optional] Are there any post deployment tasks we need to perform?
Yes the tests are not running on the github right now

## [optional] Are there any dependencies on other PRs or Work?

### CODE REVIEW

**Review by cpcundill — APPROVED** [review]

File: tests/unit/test_main.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+                "field": "geometry",
+                "issue-type": "invalid geometry - fixed",
+                "value": "Ring Self-intersection[0.124675 51.139896]",
+                "message": "None",
+            },
+        ],
+    )
+
+
+def test_bucket_and_file_existence(s3_client, s3_bucket):
```
Comment (cpcundill) [inline tests/unit/test_main.py]: Why do we need to test this?  That's not a feature of our code right?

File: tests/unit/test_main.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    assert bucket_name in bucket_names, f"Bucket {bucket_name} does not exist"
+
+    # Test if the file exists in the bucket
+    files = s3_client.list_objects(Bucket=bucket_name).get("Contents", [])
+    file_keys = [file["Key"] for file in files]
+    assert "test/path/issues.parquet" in file_keys, "File does not exist in the bucket"
+
+
+def test_duckdb_query_with_localstack(s3_client, s3_bucket, duckdb_connection):
+    # Query to fetch data from the Parquet file in S3
```
Comment (cpcundill) [inline tests/unit/test_main.py]: Again we're not  testing a feature of our code here right?

**Review by cpcundill — CHANGES_REQUESTED** [review]

File: tests/unit/test_main.py (hunk 1 above)
Comment (ssadhu-sl) [reply tests/unit/test_main.py]: No i just added some additional tests initially to make sure the bucket I am creating exists. You are right, it shouldn't be in unit tests. I ll remove them now

File: tests/integration/test_main.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+def test_search_issues(s3_bucket, duckdb_connection):
+    # Prepare test params
+    params = {
+        "dataset": "conservation-area",
+        "offset": 0,
+        "limit": 10,
+    }
+
+    # Test the function that interacts with DuckDB and S3 via LocalStack
+    with patch("db.duckdb.connect", return_value=duckdb_connection):
```
Comment (cpcundill) [inline tests/integration/test_main.py]: Why do we need to patch this method and intercept the duckdb_connection?  Ideally the integration test would just let this happen?

File: tests/integration/test_main.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        == "0b4284077da580a6daea59ee2227f9c7c55a9a45d57ef470d82418a4391ddf9a"
+    )
+
+
+def test_search_issues_no_parameters(duckdb_connection):
+    # Prepare test params
+    params = {}
+
+    # Test the function that interacts with DuckDB and S3 via LocalStack
+    with patch("db.duckdb.connect", return_value=duckdb_connection):
```
Comment (cpcundill) [inline tests/integration/test_main.py]: Again, same as above - can we not let the integration test go right through the db package which will then use DuckDb itself to connect to the S3 bucket being provided by localstack?

**Review by cpcundill — CHANGES_REQUESTED** [review]
Think we could have tests that target the db package separately to main...

File: tests/integration/test_main.py (hunk 4 above)
Comment (ssadhu-sl) [reply tests/integration/test_main.py]: If we do not mock the duck db connection, it tries to access an actual S3 endpoint which do not exist. Only when i mock it and provide the s3_endpoint as testcontainer's url, it works.

**Review by cpcundill — APPROVED** [review]

**Pull request merged**